src/Run.js

```diff
@@ -27,30 +34,46 @@ function Run(ref, json_q, callback,progressObj) {
     for (var image in json_q) {
       var no_steps = ref.images[image].steps.length;
       var init = json_q[image];
-      for(var i = 0; i < no_steps - init; i++) {
+      for (var i = 0; i < no_steps - init; i++) {
         drawarray.push({ image: image, i: init + i });
       }
     }
-    drawStep(drawarray,0);
+    drawStep(drawarray, 0);
   }
 
-  function filter(json_q){
+  function filter(json_q) {
     for (var image in json_q) {
       if (json_q[image] == 0 && ref.images[image].steps.length == 1)
         delete json_q[image];
       else if (json_q[image] == 0) json_q[image]++;
     }
     for (var image in json_q) {
       var prevstep = ref.images[image].steps[json_q[image] - 1];
-      while (typeof(prevstep) == "undefined" || typeof(prevstep.output) == "undefined") {
-        prevstep = ref.images[image].steps[(--json_q[image]) - 1];
+      while (
+        typeof prevstep == "undefined" ||
+        typeof prevstep.output == "undefined"
+      ) {
+        prevstep = ref.images[image].steps[--json_q[image] - 1];
       }
     }
     return json_q;
   }
+  function generateOutputsArray(input, i, image) {
```

Comment:
Let's have this function return the result rather than operate on the passed object? And can we name it generateStepApi () maybe?

Reply:
@jywarren yeah that makes sense, actually I didn't like this name myself😀